### ISSUE TITLE
Whitelisting DWeb Protocols

### ISSUE THREAD

**Issue by lidel** [ipfs-shipyard/arewedistributedyet#23]
## Summary

We want dweb protocols to be available without `web+` prefix in contexts such as [`navigator.registerProtocolHandler`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler) and WebExtensions ([manifest.json/protocol_handlers](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/manifest.json/protocol_handlers), if vendor supports it).

Namely: "ssb", "dat", "ipfs", "ipns", "dweb" (comment if something else should be added to the list)

## Motivation

Whitelisting protocols in above contexts is an easy change and it makes it possible for community to start using non-HTTP addresses with HTTP-based gateways/readers even before native protocol handler API is introduced.

Additionally, if we have at least two vendors implementing proposed change we would have a solid case for adding them as [safelisted-scheme in HTML Spec](https://html.spec.whatwg.org/multipage/system-state.html#safelisted-scheme).

## Resources

- [List of safelisted-schemes in HTML Spec](https://html.spec.whatwg.org/multipage/system-state.html#safelisted-scheme)
- Firefox 59 [whitelists protocol handlers for work on decentralization technologies](https://hg.mozilla.org/integration/mozilla-inbound/rev/c2cb8a06bcf1) (webextensions only)
- [[whatwg] Adding "ipfs" to the safelisted schemes](https://lists.w3.org/Archives/Public/public-whatwg-archive/2018Jul/0000.html) by @mathiasrw
  - stalled due to lack of vendors implementing change for `navigator.registerProtocolHandler`

**Comment by lidel** [comment]
@KenjiBaheux regarding "registering whitelisted dweb protocols": yes, whitelisting dweb protocols in Chrome, would be a valuable step in the right direction!

With current Firefox Stable (61)  the "user installs extension Foo and is able to open content from foo:// URLs" is possible thanks to:

- **Whitelisting dweb protocols** so we can use them without "web+" prefix
  - Firefox 59 [whitelisted dweb protocols](https://bugzilla.mozilla.org/show_bug.cgi?id=1428446) in webextension context, which means they don't need to be prefixed with "web+" and are actually usable in the wild.
   This should also be safe/easy to ship, and it would mean we have two vendors on board to push discussions like this one bit further (for all of dweb protocols).
- **Simplified handler that is only a redirect** to an HTTP page 
  - Firefox 54 implemented [redirect-based handler that is specified via extension manifest](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/protocol_handlers).
     - This is not a native handler, it is just a redirect to a HTTP page, but it makes links to non-HTTP protocol work in browser, solving basic UX of being able to access content.
     - IIUC it is basically a thin wrapper on top of [`navigator.registerProtocolHandler`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerProtocolHandler) so it should be safe to implement the same in Chrome.

Right now Chrome does not allow webextensions to automatically register basic redirect handler nor manually register dweb protocol without web+ prefix.

Would be greeat if Chrome could follow what was done in Firefox.
Having that we could start adopting URI schemes other than `https://` and be sure users are able to at least open them with WebExtensions or via HTTP gateways

**Comment by KenjiBaheux** [comment]
We are tracking the whitelisting of dweb protocols via navigator.registerProtocolHandler in [crbug.com/651311](crbug#651311). 
 - Still being discussed internally. But there is support for the dweb protocols as long as they don't have human readable parts which could be a source of confusion for unsuspecting users.
 - We would probably do this in general (i.e. not limited to a webextension context). In that case, would we still need the "simplified handler ..." item noted above?

**Comment by lidel** [comment]
@KenjiBaheux  Without "simplified handler" UX for registering support for URLs with a new protocol  is okay for websites, but still quite  bad on webextension side. Take a look at a simplified flow:
1. user installs extension to support protocol `foo://`, but it is not enough to open `foo://bar`
1. user needs to manually visit an arbitrary website (extension can open it on install, but people hate popups and close them before they load)
1. website calls `navigator.registerProtocolHandler` to register itself as redirect-based handler for the `foo://` protocol
1. browser displays [permission prompt](https://www.w3.org/TR/2013/CR-html5-20130806/images/sample-content-handler-registration.png)
1. user needs to manually accept website for handling  URLs with new protocol
1. finally, if all the above went ok, user can open `foo://bar`

This means  user has to not only install browser extension but also go over all those steps, otherwise there will be no support for new protocol.   Still, even `navigator.registerProtocolHandler` alone is much better than nothing :)

----

With very thin sugar-coating provided by webextension's [manifest.json/protocol_handlers](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/protocol_handlers) UX is much better:

> 
> The [protocol_handler](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/protocol_handlers) API lets you easily map protocols to actions in your extension. For example: we use irccloud at Mozilla, so we can map ircs:// links to irccloud by adding this into an extension:
>   ```json
>   "protocol_handlers": [
>     {
>       "protocol": "ircs",
>       "name": "IRC Mozilla Extension",
>       "uriTemplate": "https://irccloud.mozilla.com/#!/%s"
>     }
>   ]
>  ```
> 
> When a user clicks on an IRC link, it shows the application selector with the IRC Mozilla Extension visible:
 >  <img width="305" alt="screenshot-2017-03-03-14 47 52" src="https://user-images.githubusercontent.com/157609/44097165-735767d4-9fdd-11e8-8530-1378b0e30d67.png">

**Comment by KenjiBaheux** [comment]
We just posted our [public intent to implement and ship additional safelisted protocols](https://groups.google.com/a/chromium.org/forum/#!topic/blink-dev/29sFh4tTdcs). 

I don't expect any major push back but feel free to voice your support and share details about what this enables for you and end-users.

**Comment by lidel** [comment]
@KenjiBaheux I filled https://crbug.com/883274 (_Extensions API should implement manifest.json/protocol_handlers_) to address UX issues from https://github.com/arewedistributedyet/arewedistributedyet/issues/23#issuecomment-412892777. 
It is an additional task on top of  _safelisting DWeb protocols_, which is tracked in https://crbug.com/651311.

**Comment by agentofuser** [comment]
It would be great to have `git://` whitelisted as well.

**Comment by lidel** [comment]
@agentofuser  `git://` and other VCS are tracked in https://github.com/whatwg/html/pull/1829 :)

**Comment by olizilla** [comment]
Tested this and it seems that Firefox **does not** currently suppport `navigator.registerProtocolHandler` for `ipfs` and other dweb protocols, as per https://github.com/mozilla/standards-positions/issues/339

and the whatwg/html spec change is waiting on a positive response from Mozilla
> Just an update on this, the current list is
>
> cabal
> dat
> did
> dweb
> ethereum
> hyper
> ipfs
> ipns
> ssb
> 
>    - spec, WPT, mozilla and chromium patches have been updated accordingly.
>    - Chromium API owners approved this change
>    - still waiting feedback from Mozilla on mozilla/standards-positions#339
> – https://github.com/whatwg/html/pull/5482#issuecomment-644217335

**Comment by fred-wang** [comment]
Yes, and Mozilla's bug is https://bugzilla.mozilla.org/show_bug.cgi?id=1631446
